tests/fuzz/fuzz-bolt12-tlvspan.c

```diff
@@ -0,0 +1,23 @@
+#include "config.h"
+#include <common/utils.h>
+#include <wire/wire.h>
+#include <tests/fuzz/libfuzz.h>
+
+/* Include bolt12.c directly, to gain access to tlv_span(). */
+#include "../../common/bolt12.c"
+
+void init(int *argc, char ***argv)
+{}
+
+void run(const u8 *data, size_t size)
+{
+	if (size < 2 * sizeof(u64))
+		return;
+	u64 minfield = fromwire_u64(&data, &size);
+	u64 maxfield = fromwire_u64(&data, &size);
+
+	const u8 *buf = tal_dup_arr(tmpctx, u8, data, size, 0);
+	tlv_span(buf, minfield, maxfield, &size);
```

Comment:
A possible additional check:

```suggestion
	size_t span_size = tlv_span(buf, minfield, maxfield, &span_start_offset);
	assert(span_start_offset + span_size <= size);
```

Reply:
Yep, that's UB.  I'm not sure if we'd ever encounter an offer with 0 length in practice, even in an adversarial case.

But the fix seems simple enough either way.  Probably just need to change the initial value of `end`:

```c
size_t tlv_span(const u8 *tlvstream, u64 minfield, u64 maxfield,
                size_t *startp)
{
        const u8 *cursor = tlvstream;
        size_t tlvlen = tal_bytelen(tlvstream);
        const u8 *start, *end;

        start = NULL;
        end = tlvstream + tlvlen;
        while (tlvlen) {
                const u8 *before = cursor;
                bigsize_t type = fromwire_bigsize(&cursor, &tlvlen);
                bigsize_t len = fromwire_bigsize(&cursor, &tlvlen);
                if (type >= minfield && start == NULL)
                        start = before;
                if (type > maxfield)
                        break; 
                fromwire_pad(&cursor, &tlvlen, len);
                end = cursor;        
        }
        if (!start)
                start = end;

        if (startp)
                *startp = start - tlvstream;
        return end - start;
}
```

Reply:
I looked around and found only one invocation of `tlv_span()`, in `plugins/fetchinvoice.c`:
```
static bool invoice_matches_request(struct command *cmd, const u8 *invbin,
const struct tlv_invoice_request *invreq)
{
        size_t ir_len1, ir_len2, ir_start1, ir_start2;
        size_t inv_len1, inv_len2, inv_start1, inv_start2;
        u8 *wire;

        /* We linearize then strip signature.  This is dumb! */
        wire = tal_arr(tmpctx, u8, 0);
        towire_tlv_invoice_request(&wire, invreq);
        ir_len1 = tlv_span(wire, 0, 159, &ir_start1);
        ir_len2 = tlv_span(wire, 1000000000, 2999999999, &ir_start2);

        inv_len1 = tlv_span(invbin, 0, 159, &inv_start1);
        inv_len2 = tlv_span(invbin, 1000000000, 2999999999, &inv_start2);
        return memeq(wire + ir_start1, ir_len1,
                     invbin + inv_start1, inv_len1)
                && memeq(wire + ir_start2, ir_len2,
                         invbin + inv_start2, inv_len2);
}
```
And walking up this call stack:
```
static struct command_result *handle_invreq_response(struct command *cmd, struct sent *sent,
const char *buf, const jsmntok_t *om)
{
...
...
        if (!invoice_matches_request(cmd, invbin, sent->invreq)) {
                badfield = "invoice_request match";
                goto badinv;
        }
```

```
  struct command_result *handle_invoice_onion_message(struct command *cmd, const char *buf,
                                                    const jsmntok_t *om, const struct secret *pathsecret)
{
...
...
        if (sent->invreq)
                return handle_invreq_response(cmd, sent, buf, om);

        return NULL;
}
```

So it doesn't look like the undefined behavior can be triggered externally. Would we still want to guard against it though, perhaps to be future-proof?